clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3584,7 +3633,10 @@ void handleResponse(AbstractResponse abstractResponse) {
 
                 // If coordinator changed since we fetched it, retry
                 if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
-                    rescheduleFindCoordinatorTask(context, () -> getRemoveMembersFromGroupCall(context));
+                    long nextAllowedTryMs = calculateNextAllowedRetryMs();
+                    Call call = getRemoveMembersFromGroupCall(context, this.tries() + 1,
+                        nextAllowedTryMs);
+                    rescheduleFindCoordinatorTask(context, () -> call);
```

Comment:
I think there's a reasonable way to consolidate the bookkeeping in `rescheduleFindCoordinatorTask`. If we add the failed call to the argument list, then we can do something like this:
```java
    private void rescheduleFindCoordinatorTask(ConsumerGroupOperationContext<?, ?> context,
                                               Call failedCall,
                                               Supplier<Call> nextCall) {
        log.info("Node {} is no longer the Coordinator. Retrying with new coordinator.",
                context.node().orElse(null));
        // Requeue the task so that we can try with new coordinator
        context.setNode(null);

        Call call = nextCall.get();
        call.tries = failedCall.tries + 1;
        call.nextAllowedTryMs = calculateNextAllowedRetryMs();

        if (call.tries > maxRetries) {
            log.debug("Max retries {} for {} reached", maxRetries, call);
            call.fail(time.milliseconds(), new TimeoutException());
            return;
        }
        Call findCoordinatorCall = getFindCoordinatorCall(context, nextCall);
        runnable.call(findCoordinatorCall, time.milliseconds());
    }
```

Then the caller changes to this (for example):

```java
                    Call call = getDescribeConsumerGroupsCall(context);
                    rescheduleFindCoordinatorTask(context, this, () -> call);
```

What do you think?

Reply:
Yeah I think this does make it a lot cleaner and easier to read. I will refactor to change it to this, and experiment with checking the max tries in `enqueue` as well.